From daff49fd37f7c627596166a308eab812411de8ed Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 8 Mar 2022 14:11:16 +0100
Subject: [PATCH 18/22] winex11.drv: Avoid focusing windows which never were
 activated.

X11 focus fallback logic usually prefers activating windows in the
reverse order they were mapped, regardless of whether they ever were
focused before or not.

Windows which are created with the SWP_NOACTIVATE flag can be shown
without being activated, and should not be considered when the focus
changes automatically when the foreground window is closed.

CW-Bug-Id: #19553
CW-Bug-Id: #19644
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=2155
---
 dlls/winex11.drv/event.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -664,15 +664,19 @@ static void set_foreground_window( HWND hwnd )
 /**********************************************************************
  *              set_focus
  */
-static void set_focus( Display *display, HWND hwnd, Time time, BOOL check )
+static void set_focus( Display *display, Window window, HWND hwnd, Time time, BOOL check )
 {
     HWND focus, old_active;
-    Window window;
     GUITHREADINFO threadinfo;
+    Time user_time;
 
     old_active = NtUserGetForegroundWindow();
 
-    if (!(focus = find_activatable_window( hwnd, time, check ))) return;
+    /* if the window was never activated and never received user input, it should not activate on focus change */
+    if (!get_window_user_time( display, window, &user_time ) || user_time) focus = hwnd;
+    else focus = old_active;
+
+    if (!(focus = find_activatable_window( focus, time, check ))) return;
     TRACE( "setting foreground window to %p\n", focus );
     set_foreground_window( focus );
 
@@ -791,7 +795,7 @@ static void handle_wm_protocols( HWND hwnd, XClientMessageEvent *event )
                hwnd, NtUserIsWindowEnabled(hwnd), NtUserIsWindowVisible(hwnd), NtUserGetWindowLongW(hwnd, GWL_STYLE),
                get_focus(), get_active_window(), NtUserGetForegroundWindow(), last_focus );
 
-        set_focus( event->display, hwnd, event_time, TRUE );
+        set_focus( event->display, event->window, hwnd, event_time, TRUE );
     }
     else if (protocol == x11drv_atom(_NET_WM_PING))
     {
@@ -865,7 +869,7 @@ static BOOL X11DRV_FocusIn( HWND hwnd, XEvent *xev )
         return TRUE;
     }
 
-    set_focus( event->display, hwnd, CurrentTime, FALSE );
+    set_focus( event->display, event->window, hwnd, CurrentTime, FALSE );
     return TRUE;
 }
 
-- 
2.36.1

